File: lintcode/best-time-to-buy-and-sell-stock-ii.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/best-time-to-buy-and-sell-stock-ii/

class Solution:
    """
    @param prices: Given an integer array
    @return: Maximum profit
    """
    def maxProfit(self, prices):
        # write your code here
        result = 0
        i = 1
        while i < len(prices):
            while i < len(prices) and prices[i] <= prices[i-1]:
                i += 1

            j = i
            while j < len(prices) and prices[j] >= prices[j-1]:
                j += 1

            result += prices[j-1] - prices[i-1]
            i = j

        return result

if __name__ == '__main__':
    s = Solution()
    print s.maxProfit([2,1,2,0,1])

File: leetcode/best-time-to-buy-and-sell-stock-ii.cc
// Question: https://leetcode.com/problems/best-time-to-buy-and-sell-stock-ii/

#include "headers.h"

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if (prices.empty()) {
            return 0;
        }

        int ans = 0;
        int min_price = prices[0];
        int i = 1;
        int len = prices.size();
        while (i < len) {
            int j = i;
            while (i < len && prices[i] >= prices[i-1]) {
                ++i;
            }
            ans += prices[i-1] - min_price;
            if (j == i) {
                ++i;
            }

            if (i == len) {
                break;
            }
            min_price = prices[i-1];
        }

        return ans;
    }
};

int main(int argc, char** argv) {
    vector<int> nums{1,4,2};
    Solution s;
    cout << s.maxProfit(nums) << endl;
    return 0;
}
